public/app/app.ts

```diff
@@ -380,4 +381,49 @@ function reportMetricPerformanceMark(metricName: string, prefix = '', suffix = '
   }
 }
 
+function handleRedirectTo(): void {
+  const queryParams = new URLSearchParams(window.location.search);
+  const redirectToParamKey = 'redirectTo';
+
+  if (queryParams.has(redirectToParamKey) && window.location.pathname !== '/') {
+    const rawRedirectTo = queryParams.get(redirectToParamKey)!;
+    if (!isValidRedirectTo(rawRedirectTo)) {
+      return;
+    }
+    window.sessionStorage.setItem(RedirectToUrlKey, decodeURIComponent(rawRedirectTo));
+    queryParams.delete(redirectToParamKey);
+    window.history.replaceState({}, '', `${window.location.pathname}${queryParams.size > 0 ? `?${queryParams}` : ''}`);
+    return;
+  }
+
+  const redirectTo = window.sessionStorage.getItem(RedirectToUrlKey);
+  if (!redirectTo) {
+    return;
+  }
+
+  if (!isValidRedirectTo(redirectTo)) {
+    window.sessionStorage.removeItem(RedirectToUrlKey);
+    return;
+  }
+
+  window.sessionStorage.removeItem(RedirectToUrlKey);
+  window.location.replace(redirectTo);
+}
+
+function isValidRedirectTo(path: string): boolean {
+  if (!path.startsWith('/')) {
+    return false;
+  }
+
+  if (path.startsWith('//')) {
+    return false;
+  }
+
+  if (config.appSubUrl && !path.startsWith(config.appSubUrl) + '/') {
+    return false;
+  }
+
+  return true;
+}
```

Comment:
Hmm. Perhaps we should always enforce that we redirect to a path of the origin?

Something like

```js
const redirectParam = sessionStorage.getItem(...);
const redirectTo = window.location.origin + redirectParam;
window.location.replace(redirectTo);
```

?